### PR TITLE
OPENEUROPA-1378: Implement DateValueObject for Date pattern fields

### PR DESCRIPTION
**Pull request by drupol**

### CODE REVIEW

File: templates/patterns/date_block/date_block.ui_patterns.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      type: "string"
+      label: "Month"
+      description: "(optional) Numeric or textual representation of a month."
+      preview: "9"
+    year:
+      type: "integer"
+      label: "Year"
+      description: "(optional) A full numeric representation of a year, 4 digits."
+      preview: "2018"
+    timestamp:
```
Comment (imanoleguskiza) [inline templates/patterns/date_block/date_block.ui_patterns.yml]: Why did you add this parameter and added all the extra functionality? ECL doesn't mention it and I don't see it as a requirement in the ticket

File: src/ValueObject/DateValueObject.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+   * @param string $day
+   *   The date day.
+   * @param string $month
+   *   The date month.
+   * @param string $year
+   *   The date year.
+   * @param string $variant
+   *   The pattern variant.
+   */
+  private function __construct(string $day, string $month, string $year, string $variant = 'default') {
```
Comment (ademarco) [inline src/ValueObject/DateValueObject.php]: I'd keep the current approach with setters and getters for now, we have a follow up issue to consider the alternative proposed by @drupol , until that's not discussed and agreed we should not change direction. Refactorying would be relatively quick anyway and can be assigned to a person different than @drupol so to spread more the knowledge about the new approach, if accepted.

File: tests/Unit/ValueObject/DateValueObjectTest.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Test date value object.
+ */
+class DateValueObjectTest extends UnitTestCase {
+
+  /**
+   * Test constructing a date value object from an array.
+   */
+  public function testFromArray() {
```
Comment (imanoleguskiza) [inline tests/Unit/ValueObject/DateValueObjectTest.php]: If we are testing the fromArray method I guess testing the other methods (fromTimestamp, from DateTimePlus) is also a good idea

File: tests/fixtures/rendering.yml [hunk 4]
```diff
@@ -1,3 +1,59 @@
+- array:
```
Comment (imanoleguskiza) [inline tests/fixtures/rendering.yml]: We only have rendering tests that use the timestamp, in theory it should be possible to pass the individual variables and so I think we should test it as well.

File: templates/patterns/date_block/date_block.ui_patterns.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      type: "integer"
+      label: "Timestamp"
+      description: "(optional) A full numeric representation of a year, 4 digits. When this is set, it will override other parameters."
+      preview: "1543583805"
```
Comment (ademarco) [inline templates/patterns/date_block/date_block.ui_patterns.yml]: We should have here only one field called "date" and of "type" DateValueObject, then in the preprocess we build the value object.

File: templates/patterns/date_block/date_block.ui_patterns.yml (hunk 1 above)
Comment (drupol) [reply templates/patterns/date_block/date_block.ui_patterns.yml]: Fixed.

File: tests/Unit/ValueObject/DateValueObjectTest.php (hunk 3 above)
Comment (drupol) [reply tests/Unit/ValueObject/DateValueObjectTest.php]: Ok

File: templates/patterns/date_block/date_block.ui_patterns.yml (hunk 5 above)
Comment (drupol) [reply templates/patterns/date_block/date_block.ui_patterns.yml]: Ok

File: src/ValueObject/DateValueObject.php (hunk 2 above)
Comment (drupol) [reply src/ValueObject/DateValueObject.php]: See #151

File: tests/fixtures/rendering.yml (hunk 4 above)
Comment (drupol) [reply tests/fixtures/rendering.yml]: Done.

File: tests/Unit/ValueObject/DateValueObjectTest.php (hunk 3 above)
Comment (drupol) [reply tests/Unit/ValueObject/DateValueObjectTest.php]: I removed the methods from the DateValueObject because they are no more relevant for now.

File: src/ValueObject/DateValueObject.php (hunk 2 above)
Comment (drupol) [reply src/ValueObject/DateValueObject.php]: @ademarco I'm using variables in the constructor, nothing else for now... I guess this comment is outdated then.
I removed the "withers" as well.

File: tests/fixtures/rendering.yml [hunk 6]
```diff
@@ -321,27 +321,6 @@
     equals:
       'h3.ecl-list-item__title': "Item title"
       'div.ecl-list-item__detail': "Item text"
-- array:
```
Comment (imanoleguskiza) [inline tests/fixtures/rendering.yml]: Why was this test removed?

File: tests/Unit/ValueObject/DateValueObjectTest.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
```
Comment (imanoleguskiza) [inline tests/Unit/ValueObject/DateValueObjectTest.php]: Do we need to keep this test if we already have the DateBlockPatternRenderingTest.php? If so, we still need a test for the fromTimestamp method here

File: tests/fixtures/rendering.yml (hunk 6 above)
Comment (drupol) [reply tests/fixtures/rendering.yml]: As the pattern is using a `DateValueObject`, we need to deal with an object, not only an array.
This is why this test was removed from there and moved into `DateBlockPatternRenderingTest.php`

File: tests/Kernel/Patterns/DateBlockPatternRenderingTest.php [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  }
+
+  /**
+   * Data provider for testFromTimestamp.
+   *
+   * @return array
+   *   An array of test data arrays with assertions.
+   */
+  public function fromTimestampDataProvider(): array {
+    return [
```
Comment (voidtek) [inline tests/Kernel/Patterns/DateBlockPatternRenderingTest.php]: Why we are not using an external yml file, as `renderingDataProvider`?

File: src/ValueObject/DateValueObject.php [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+   * @param string $year
+   *   The date year.
+   * @param string|int|null $weekDay
+   *   The day of the week.
+   */
+  private function __construct(string $day, string $month, string $year, string $weekDay = NULL) {
+    $this->day = $day;
+    $this->month = $month;
+    $this->year = $year;
+    $this->weekDay = $weekDay;
```
Comment (voidtek) [inline src/ValueObject/DateValueObject.php]: This line is useless, duplicate of  
```
$this->weekDay = empty($weekDay) ?
      $date->format('l') :
      $weekDay;
```

File: src/ValueObject/DateValueObject.php [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+    $date = new DateTimePlus(implode('-', [$year, $month, $day]));
+
+    $this->weekDay = empty($weekDay) ?
+      $date->format('l') :
+      $weekDay;
+    $this->monthName = $date->format('F');
```
Comment (voidtek) [inline src/ValueObject/DateValueObject.php]: I am not comfortable to have this `monthName` created only by the `DateTimePlus` and not an override if it's `NULL`

File: src/ValueObject/DateValueObject.php [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function fromArray(array $parameters = []): ValueObjectInterface {
+    return new static(
+      $parameters['day'],
+      $parameters['month'],
+      $parameters['year']
```
Comment (voidtek) [inline src/ValueObject/DateValueObject.php]: the parameter `weekDay` is missing.

File: templates/patterns/date_block/date_block.ui_patterns.yml [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+      description: "Past event"
+  fields:
+    date:
+      type: "DateValueObject"
+      label: "Date"
+      description: "A date value object."
+      preview:
+        day: "24"
+        month: "09"
+        year: "1981"
```
Comment (voidtek) [inline templates/patterns/date_block/date_block.ui_patterns.yml]: we should add the parameter `weekday: "Thu."`

**Comment by ademarco** [comment]
After discussing it with @voidtek we realized that it would be simpler if we just construct the date using timestamps, if we do so we can easily accommodate date periods, which is actually in line with ECL specifications.

File: tests/Kernel/fixtures/patterns/date_block_pattern_rendering.yml [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+      div.ecl-date-block.ecl-date-block--default: 1
+      div.ecl-date-block.ecl-date-block--cancelled: 0
+      div.ecl-date-block.ecl-date-block--past: 0
+    equals:
+      span.ecl-date-block__week-day: Thu
+      span.ecl-date-block__day: 24
+      span.ecl-date-block__month: Sep
+      span.ecl-date-block__year: 1981
+
+# Test "cancelled" variant.
```
Comment (voidtek) [inline tests/Kernel/fixtures/patterns/date_block_pattern_rendering.yml]: We forget to check the variant "ecl-date-block ecl-date-block--ongoing"

File: src/ValueObject/DateValueObject.php [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+      $this->end = DateTimePlus::createFromTimestamp($end, $timezone);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function fromArray(array $parameters = []): ValueObjectInterface {
+    return new static(
+      $parameters['start'],
```
Comment (voidtek) [inline src/ValueObject/DateValueObject.php]: We should check if $parameters['start'] exists.

File: src/ValueObject/DateValueObject.php [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+    if ($end !== NULL) {
+      $this->end = DateTimePlus::createFromTimestamp($end, $timezone);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function fromArray(array $parameters = []): ValueObjectInterface {
```
Comment (voidtek) [inline src/ValueObject/DateValueObject.php]: `ValueObjectInterface` should be `DateValueObjectInterface`

File: src/ValueObject/DateValueObject.php [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+   *   Format to be used to print the interval.
+   * @param string $extra
+   *   Run extra check to make sure we should not print an interval.
+   *   This is useful, for example, in case of same days on different months
+   *   which should still be printed as an interval.
+   *
+   * @return string
+   *   Formatted interval.
+   */
+  protected function getDateInterval(string $format, string $extra): string {
```
Comment (voidtek) [inline src/ValueObject/DateValueObject.php]: We should improve this function by putting variables, to be consistent.

File: src/ValueObject/DateValueObjectInterface.php [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+   *   Start date as UNIX timestamp.
+   * @param int $end
+   *   End date as UNIX timestamp.
+   * @param string|null $timezone
+   *   Timezone string, e.g. "Europe/Brussels".
+   *
+   * @return \Drupal\oe_theme\ValueObject\DateValueObjectInterface
+   *   A new DateValueObject.
+   */
+  public static function fromTimestamp(int $start, int $end = NULL, string $timezone = NULL): DateValueObjectInterface;
```
Comment (voidtek) [inline src/ValueObject/DateValueObjectInterface.php]: The function `fromArray ` is missing.

File: src/ValueObject/DateValueObject.php (hunk 14 above)
Comment (drupol) [reply src/ValueObject/DateValueObject.php]: Fixed in upcoming commit.

File: src/ValueObject/DateValueObject.php (hunk 15 above)
Comment (drupol) [reply src/ValueObject/DateValueObject.php]: No, you cannot override what's in the Interface.

File: tests/Kernel/fixtures/patterns/date_block_pattern_rendering.yml (hunk 13 above)
Comment (drupol) [reply tests/Kernel/fixtures/patterns/date_block_pattern_rendering.yml]: Fixed.

File: tests/Kernel/fixtures/patterns/date_block_pattern_rendering.yml [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+      div.ecl-date-block.ecl-date-block--past: 1
+      div.ecl-date-block.ecl-date-block--ongoing: 0
+    equals:
+      span.ecl-date-block__week-day: Thu
+      span.ecl-date-block__day: 24
+      span.ecl-date-block__month: Sep
+      span.ecl-date-block__year: 1981
+
+# Test "unknown" variant.
+- variant: unexistant
```
Comment (voidtek) [inline tests/Kernel/fixtures/patterns/date_block_pattern_rendering.yml]: This variant is out of scope and I don't see the purpose.
This comment is not a blocker.

Comment (drupol) [reply tests/Kernel/fixtures/patterns/date_block_pattern_rendering.yml]: I think this should be part of the test, to check if when rendering a variant that is unknown, it doesn't set any specific existing classes.

Comment (ademarco) [reply tests/Kernel/fixtures/patterns/date_block_pattern_rendering.yml]: I think it's good to test what happen when the variant is either empty or not a valid one, we should then fallback to the default variant. We can add this logic directly in the pattern template or in the global pattern preprocess.

Comment (drupol) [reply tests/Kernel/fixtures/patterns/date_block_pattern_rendering.yml]: Fixed!

File: tests/Kernel/Patterns/DateBlockPatternRenderingTest.php (hunk 8 above)
Comment (drupol) [reply tests/Kernel/Patterns/DateBlockPatternRenderingTest.php]: Same reason as previously explained.

File: src/ValueObject/DateValueObject.php (hunk 9 above)
Comment (drupol) [reply src/ValueObject/DateValueObject.php]: Fixed.

File: src/ValueObject/DateValueObject.php (hunk 10 above)
Comment (drupol) [reply src/ValueObject/DateValueObject.php]: Fixed.

File: src/ValueObject/DateValueObject.php (hunk 11 above)
Comment (drupol) [reply src/ValueObject/DateValueObject.php]: Fixed.

File: templates/patterns/date_block/date_block.ui_patterns.yml (hunk 12 above)
Comment (drupol) [reply templates/patterns/date_block/date_block.ui_patterns.yml]: Fixed.

File: oe_theme.theme [hunk 19]
```diff
@@ -644,6 +641,26 @@ function oe_theme_preprocess_paragraph__oe_content_row__variant_inpage(array &$v
   $variables['links'] = $links;
 }
 
+/**
+ * Implements hook_preprocess_pattern().
+ */
+function oe_theme_preprocess_pattern_date_block(array &$variables): void {
+  // This make sure that the variant is not empty and has a "default" value.
+  // @todo: How about doing this in all components ?
+  // @todo: We should optimize how to compute the list of variants.
```
Comment (ademarco) [inline oe_theme.theme]: Comment should rather be:
```
// This make sure that the variant value will always fallback to "default" if an empty or not valid value is passed.
```

I'd also remove the todo here since it's not something to do about the current block of code, I'd rather open a follow up to do this on all components.

Comment (drupol) [reply oe_theme.theme]: Updating now.

File: docs/developer-documentation.md [hunk 20]
```diff
@@ -85,6 +85,24 @@ Check the [UI Patterns documentation][2] for more information about how to use p
...
 
+#### `DateValueObject`
+
+Used in the following patterns:
+
+- [`date_block`](../templates/patterns/date_block/date_block.ui_patterns.yml)
+
+Provides the following factory methods:
+
+- `DateValueObject::fromArray(array $values = [])`: accepts an array with the following properties:
```
Comment (imanoleguskiza) [inline docs/developer-documentation.md]: The PR is fine, but do we really need this static method? Its essentially the same as the fromTimestamp method but a bit worse since fromTimestamp tells you what parameters to pass.

Comment (drupol) [reply docs/developer-documentation.md]: We need this, yes. It's defined in the the interface... and we cannot go over it.
